### ISSUE TITLE
Merchant in refugee camp do not restock 

### ISSUE THREAD

**Issue by ynemo** [CleverRaven/Cataclysm-DDA#46766]
1st time i made deal with Main merchant in ref. camp, i bought few bullets with the change.
Then i realized there is full 50 stack of bullets in my pocket, and there was mistake in game. 
Few days later i bought all bullets and all goods i need. I come back sometime, and now more then 40 days later, there are same goods for sale,no restock. 

# Expected behavior
They said vendor restocks every 3 days. 

- OS: Windows
    - OS Version: 10.0 1909
- Game Version: 0.E-8965-gad0e7ae-dirty [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food]

**Comment by PaulBlay** [comment]
Save game file? Also, have you tried waiting three more days and trying again.

**Comment by ynemo** [comment]
just did it, wait 3+ more days, no restock 

i keep my notes from day 6, when i made 1st deal with vendor
![image](https://user-images.githubusercontent.com/24881844/104750086-a914f080-5764-11eb-940c-4492876d480f.png)

there are 2 saves with more than 3 days difference 
[save.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/5821609/save.zip)

**Comment by mlangsdorf** [comment]
I can't reproduce on 0.E-3-g98e7362d (build 11348+)

Day 1 inventory
![Day1Inventory](https://user-images.githubusercontent.com/8007405/104753185-40eeec00-571d-11eb-9233-6b0ed5402f30.png)

Day 4 inventory
![Day4Inventory](https://user-images.githubusercontent.com/8007405/104753197-464c3680-571d-11eb-8fa4-3e3dce476d6d.png)

There's overlap that may fool a casual check, but they're clearly different inventories.

**Comment by PaulBlay** [comment]
@mlangsdorf Are those two screenshots from the two save files the OP gave?

tl;dr I checked, and it does seem to be bugged. 

- OS: Windows
    - OS Version: MINGW/CYGWIN/MSYS2 on unknown Windows version
- Game Version: 0.E-8914-g8c0034b-dirty [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food]
]

I loaded the first and cleared the content of the Merchant's inventory. I basically stole everything and used debug to remove every item he was wearing / carrying. Then I went away and hung out at a farm for four days. Then I teleported back and checked his inventory. It was completely empty.

I'll put up a save file from just before I talk to the merchant after being away.

**Comment by Cyprex** [comment]
It would likely help to add debug logging when a merchant restocks.

**Comment by PaulBlay** [comment]
Given that someone else has found a similar bug (or the same bug on a different trader) maybe the 'need confirmation' flag needs to be reviewed.

**Comment by serveral1** [comment]
From my experience (Hub01 merchant) it seems that the merchant restock algorithm just keeps endlessly spawning things on their tile until it reaches a point just where there just isn't enough volume to display the newer items being "restocked". 
I've come to this conclusion after trading multiple times with the Hub01 merchant and after some 4-5 restocks, their inventory just seemed to remain static. However, whenever I bought something from them, finished the trade and opened up the trade menu again, there would be a couple new items available (depending on the volume of the items I bought), meaning old stock is just taking over the space of the new stock (in the trade menu at least).

**Comment by PaulBlay** [comment]
> From my experience (Hub01 merchant) it seems that the merchant restock algorithm just keeps endlessly spawning things on their tile until it reaches a point just where there just isn't enough volume to display the newer items being "restocked".

That may well also happen, but in the test case I tried above I removed literally everything he had - on him or on the floor. It still hadn't restocked at all after four days.

**Comment by hexagonrecursion** [comment]
Could not reproduce. I loaded  https://github.com/CleverRaven/Cataclysm-DDA/files/5822147/EmptyShop.zip and the merchant restocked within seconds of loading the save

- OS: Linux
    - OS Version: LSB Version: :core-4.1-amd64:core-4.1-noarch; Distributor ID: Fedora; Description: Fedora release 33 (Thirty Three); Release: 33; Codename: ThirtyThree; 
- Game Version: b34a0f5 [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food]
]

**Comment by wapcaplet** [comment]
I'm gonna get ahead of stalebot and close this as unreproducible. Please ping me if anyone can still reliably reproduce it.

**Comment by stubkan** [comment]
I believe this may be an ongoing bug.  

I think what several1 is saying is happening here.  The merchants seem to not restock - because they are full of inventory that has been sold to them - ie, stuff that we do not want, because we sold it to them.  So they accumulate it, and it blocks out the restocked goods, as the list is too long.  So it appears like they aren't restocking.

You should be able to repdroduce this by selling them a lot of trash and then waiting for the restock.

Here is my game

![image](https://user-images.githubusercontent.com/34361592/166319194-70a1983e-7fc0-4f48-9f6e-2404004ff1d9.png)

The merchant area is filled completely with the stuff I sold him weeks ago

![image](https://user-images.githubusercontent.com/34361592/166319378-bd5b8435-5a42-467f-922b-25c1da2f2e8a.png)

I found a spot in the pantry that was filling up with items which looked like it was the actual restock, this is the stuff that should be sold, but isn't - it ends up outside his shop.

![image](https://user-images.githubusercontent.com/34361592/166319630-937c57f2-61cc-4f75-baff-5d8fa0b68496.png)

Is it intentional that the merchant does not delete the stuff the player sells them during a restock?  This seems like something that should be occuring, and would prevent this bug

**Comment by stubkan** [comment]
@wapcaplet

**Comment by stubkan** [comment]
Possible solutions...

Delete old items, by clearing the space around the trader.

Create infinite storage furniture for the traders use only.

Rewrite restock code to mark the blocks that recieve trader items, and to look at the marked blocks during trade panel population, to fetch them wherever they ended up getting placed.

**Comment by stubkan** [comment]
Further testing narrowed down what is happening.

There are loot zones - the merchant can access anything on these loot zones.  The stock overflow is going to those loot zones.  However it seems that the merchant can only access loot zones on the same overmap 24x24 grid that it is on.

I wrote a loot zone for the spot in the pantry that was recieving surplus items to ensure there was a loot zone - but there appears to be one already.

When the merchant moves out of his room but remains in the same original 24x24 overmap tile, he can still access the same stock.  But when he is moved to the pantry, he loses the ability to access original stock, but gains ability to access the surplus stock.

![image](https://user-images.githubusercontent.com/34361592/166342583-f916e47e-c3b3-4c7f-8b3f-7859ffc96968.png)

**Comment by andrei8l** [comment]
> When the merchant moves out of his room but remains in the same original 24x24 overmap tile, he can still access the same stock. But when he is moved to the pantry, he loses the ability to access original stock, but gains ability to access the surplus stock.

This part is intentional. Merchants can only sell items in `PICKUP_RANGE`.

> So when you purchase enough to clear space, the overflow will get moved

Items that don't fit in the loot zones get discarded so there shouldn't be any overflow. I can't (neatly) delete overflow from old saves though.

I tried reproducing your scenario by filling the spots around a merchant in an old save with 200L barrels and sticks. There was no overflow and items were consumed appropriately after passing time (in 7-days increments).

Please open a new issue with your version info and save attached.
